overlord/snapstate/handlers.go

```diff
@@ -115,6 +115,79 @@ func snapSetupAndState(t *state.Task) (*SnapSetup, *SnapState, error) {
 
 */
 
+const defaultCoreSnapName = "core"
+const defaultBaseSnapsChannel = "stable"
+
+// timeout for tasks to check if the prerequisites are ready
+var prerequisitesRetryTimeout = 1 * time.Minute
+
+func (m *SnapManager) doPrerequisites(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	// check if core is installed already
+	_, err := CoreInfo(st)
+	if err == nil {
+		return nil
+	}
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	// check if we need to inject tasks to install core
+	snapsup, _, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	snapName := snapsup.Name()
+	if snapName == defaultCoreSnapName || snapName == "ubuntu-core" {
+		return nil
+	}
+
+	// check that there is no task that installs core already
+	for _, chg := range st.Changes() {
+		if chg.Status().Ready() || chg.ID() == t.Change().ID() {
+			continue
+		}
+		for _, tc := range chg.Tasks() {
+			if tc.Kind() == "link-snap" {
+				snapsup, err := TaskSnapSetup(tc)
+				if err != nil {
+					return err
+				}
+				// some other change aleady installs core
+				if snapsup.Name() == defaultCoreSnapName {
+					// if something else installs core
+					// already we need to wait
+					return &state.Retry{After: prerequisitesRetryTimeout}
+				}
+			}
+		}
+	}
+
+	// not installed, nor queued for install -> install it
+	ts, err := Install(st, defaultCoreSnapName, defaultBaseSnapsChannel, snap.R(0), snapsup.UserID, Flags{})
+	// something might have triggered an explicit install of core while
+	// the state was unlocked -> deal with that here
+	if _, ok := err.(changeDuringInstallError); ok {
+		return &state.Retry{After: prerequisitesRetryTimeout}
+	}
+	if err != nil {
+		return err
+	}
+	ts.JoinLane(st.NewLane())
+
+	// inject install for core into this change
+	chg := t.Change()
+	for _, t := range chg.Tasks() {
+		t.WaitAll(ts)
+	}
+	chg.AddAll(ts)
```

Comment:
we need to set the task state to done directly like we do for link-snap, otherwise there's non-zero chance that we crash in the middle of exiting the task and we will have the new tasks in the change but will rerun prerequisites again and it will conflict with itself, or we need to deal with that situation otherwise